### PR TITLE
[ACIX-424] Add codeowners linter for .gitlab files

### PR DESCRIPTION
**Pull request by CelianR**
### What does this PR do?

> [!NOTE]
> Check out the [documentation](https://datadoghq.atlassian.net/wiki/spaces/ADX/pages/4180118144/Gitlab+CI+Configuration+Linters#Job-Owners-(gitlab-ci-jobs-owners%2C-gitlab-ci-jobs-codeowners)) 📖 

This adds a linter that verifies that each file within `.gitlab` has a specific codeowner rule, i.e. there are `/.gitlab/* <team>` rules within CODEOWNERS.
Only changed / new files are verified

### CODE REVIEW

File: tasks/libs/common/git.py [hunk 1]
```diff
@@ -45,6 +45,40 @@ def get_staged_files(ctx, commit="HEAD", include_deleted_files=False) -> Iterabl
...
+    - removed: Include removed files
+    - only_names: Return only the file names without the status
+    - Returns a list of (status, filename)
+    """
+
+    last_main_commit = ctx.run(f"git merge-base HEAD origin/{base_branch}", hide=True).stdout.strip()
+
+    modifications = [
+        line.split() for line in ctx.run(f"git diff --name-status {last_main_commit}", hide=True).stdout.splitlines()
+    ]
```
Comment (chouetz) [inline tasks/libs/common/git.py:67]: What about using [this](https://github.com/DataDog/datadog-agent/blob/main/tasks/libs/common/git.py#L48)?

File: tasks/libs/common/git.py [hunk 2]
```diff
@@ -45,6 +45,40 @@ def get_staged_files(ctx, commit="HEAD", include_deleted_files=False) -> Iterabl
                 yield file
 
 
+def get_file_modifications(
+    ctx, base_branch=DEFAULT_BRANCH, added=False, modified=False, removed=False, only_names=False
```
Comment (KevinFairise2) [inline tasks/libs/common/git.py:49]: That will probably not work as we expect it on release branches. Maybe you could add something like that to make sure we have the correct branch on release branches? https://github.com/DataDog/datadog-agent/blob/b7383882501637956482b60e1e112963664e74a8/tasks/gotest.py#L830

File: tasks/libs/common/git.py [hunk 3]
```diff
@@ -45,6 +45,40 @@ def get_staged_files(ctx, commit="HEAD", include_deleted_files=False) -> Iterabl
...
+    """
+    Get file status changes for the current branch compared to the base branch.
+
+    If no filter is provided, will return all the files.
+
+    - added: Include added files
+    - modified: Include modified files
+    - removed: Include removed files
+    - only_names: Return only the file names without the status
+    - Returns a list of (status, filename)
```
Comment (chouetz) [inline tasks/libs/common/git.py:60]: 🥜 We could probably use a more uniform layout in our [docstring](https://peps.python.org/pep-0257/#multi-line-docstrings)

**Review by KevinFairise2 — APPROVED** [review]

File: tasks/linter.py [hunk 4]
```diff
@@ -783,3 +784,42 @@ def gitlab_ci_jobs_owners(_, diff_file=None, config_file=None, path_jobowners='.
         )
     else:
         print(f'{color_message("Success", Color.GREEN)}: All jobs have owners defined in {path_jobowners}')
+
+
+@task
+def gitlab_ci_jobs_codeowners(ctx, path_codeowners='.github/CODEOWNERS', all_files=False):
```
Comment (chouetz) [inline tasks/linter.py:790]: Is there a reason to focus only on `gitlab` files, and not everything within the git repository?

File: tasks/linter.py [hunk 5]
```diff
@@ -783,3 +784,42 @@ def gitlab_ci_jobs_owners(_, diff_file=None, config_file=None, path_jobowners='.
...
+    """
+    Verifies that added / modified job files are defined within CODEOWNERS.
+    """
+
+    if all_files:
+        modified = glob('.gitlab/**/*.yml', recursive=True)
+    else:
+        modified = get_file_modifications(ctx, added=True, modified=True, only_names=True)
+        modified = [path for path in modified if path.endswith('.yml')]
```
Comment (chouetz) [inline tasks/linter.py:799]:
```suggestion
        modified_yml = [path for path in modified if path.endswith('.yml')]
```
🥜

File: tasks/linter.py [hunk 6]
```diff
@@ -783,3 +784,42 @@ def gitlab_ci_jobs_owners(_, diff_file=None, config_file=None, path_jobowners='.
...
+
+    with open(path_codeowners) as f:
+        lines = f.readlines()
```
Comment (chouetz) [inline tasks/linter.py:802]:
```suggestion
        parsed_owners = f.readlines()
```
You could also use [this](https://github.com/DataDog/datadog-agent/blob/main/tasks/libs/owners/parsing.py#L7)

File: tasks/linter.py [hunk 7]
```diff
@@ -783,3 +784,42 @@ def gitlab_ci_jobs_owners(_, diff_file=None, config_file=None, path_jobowners='.
...
+
+    # Keep only gitlab related lines to avoid defaults
+    lines = [line for line in lines if '/.gitlab/' in line]
+    owners = CodeOwners('\n'.join(lines))
```
Comment (chouetz) [inline tasks/linter.py:806]:
```suggestion
    gitlab_owners = CodeOwners('\n'.join(lines))
```
🥜

File: tasks/linter.py [hunk 8]
```diff
@@ -783,3 +784,42 @@ def gitlab_ci_jobs_owners(_, diff_file=None, config_file=None, path_jobowners='.
...
+
+    if not modified:
```
Comment (chouetz) [inline tasks/linter.py:808]: You can do this before loading the `CODEOWNERS`, wdyt?

**Review by chouetz — APPROVED** [review]
Only minor comments, good job

File: tasks/libs/common/git.py (hunk 1 above)
Comment (CelianR) [reply tasks/libs/common/git.py:67]: The problem with this function is that it doesn't return the status nor the renamed files. I will update my function such that `get_modified_files` will rely on it

File: tasks/linter.py (hunk 4 above)
Comment (CelianR) [reply tasks/linter.py:790]: This linter verifies that each job file has a **specific** codeowner rule. In other words, it excludes rules not specific enough such as `.*` or `*` and verifies `/.gitlab/*` patterns. Maybe you are interested in this [PR](https://github.com/DataDog/datadog-agent/pull/29695) for more general usage

File: tasks/linter.py (hunk 6 above)
Comment (CelianR) [reply tasks/linter.py:802]: As some preprocessing is made and only using `CodeOwners(<content>)` from this function is necessary, adding the preprocessing within this function is less readable IMO

File: tasks/libs/common/git.py (hunk 3 above)
Comment (CelianR) [reply tasks/libs/common/git.py]: *Taking notes*

**Pull request merged**